sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,169 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"errors"
+	"os"
+	"path"
+
+	"github.com/snapcore/snapd/sandbox/cgroup/inotify"
+)
+
+type appData struct {
+	name        string
+	cgroupPaths []string
+	npaths      int
+	channel     chan string
+}
+
+// This class allows to monitor a CGroup and detect when all the running
+// instances have been closed
+
+type CGroupMonitor struct {
+	watched map[string][]*appData
+	watcher *inotify.Watcher
+	channel chan appData
+}
+
+var currentCGroupMonitor = CGroupMonitor{
+	watcher: nil,
+	channel: make(chan appData),
+	watched: make(map[string][]*appData),
+}
+
+func deletedFile(filename string) {
+	basePath := path.Dir(filename)
+	entry := currentCGroupMonitor.watched[basePath]
+	var newList []*appData
+	for _, app := range entry {
+		for _, folder := range app.cgroupPaths {
+			if folder == filename {
+				app.npaths--
+			}
+		}
+		if app.npaths == 0 {
+			// all the folders have disappeared, so notify that this app has no more instances running
+			app.channel <- app.name
+		} else {
+			if app.npaths > 0 {
+				newList = append(newList, app)
+			}
+		}
+	}
+	if len(newList) != 0 {
+		currentCGroupMonitor.watched[basePath] = newList
+	} else {
+		delete(currentCGroupMonitor.watched, basePath)
+		currentCGroupMonitor.watcher.RemoveWatch(basePath)
+	}
+}
+
+func addFiles(newApp *appData) {
+	if newApp.npaths == 0 {
+		newApp.channel <- newApp.name
+	} else {
+		addedPaths := false
+		for _, fullPath := range newApp.cgroupPaths {
+			basePath := path.Dir(fullPath) // Monitor the path containing this folder
+			_, exists := currentCGroupMonitor.watched[basePath]
+			if !exists {
```

Comment:
reverse this check and we can remove one level of indentation
```suggestion
			if exists {
			        continue
			}
```

Reply:
Done.